maskrcnn_benchmark/structures/segmentation_mask.py

```diff
@@ -1,51 +1,93 @@
 
 # transpose
 FLIP_LEFT_RIGHT = 0
 FLIP_TOP_BOTTOM = 1
 
-
 class Mask(object):
     """
     This class is unfinished and not meant for use yet
     It is supposed to contain the mask for an object as
     a 2d tensor
     """
-
-    def __init__(self, masks, size, mode):
-        self.masks = masks
+    def __init__(self, segm, size, mode):
+        width, height = size
+        if isinstance(segm, Mask):
+            mask = segm.mask
+        else:
+            if type(segm) == list:
+                # polygons
+                rle = mask_utils.frPyObjects(segm, height, width)
+                mask = np.array(mask_utils.decode(rle), dtype=np.float32)
+                mask = np.sum(mask, axis=2)
+                mask = torch.from_numpy(np.array(mask > 0, dtype=np.float32))
+            elif type(segm) == dict and 'counts' in segm:
+                if type(segm['counts']) == list:
+                    # uncompressed RLE
+                    h, w = segm['size']
+                    rle = mask_utils.frPyObjects(segm, h, w)
+                    mask = mask_utils.decode(rle)
+                    mask = torch.from_numpy(mask).to(dtype=torch.float32)
+                else:
+                    # compressed RLE
+                    mask = mask_utils.decode(segm)
+                    mask = torch.from_numpy(mask).to(dtype=torch.float32)
+            else:
+                # binary mask
+                if type(segm) == np.ndarray:
+                    mask = torch.from_numpy(segm).to(dtype=torch.float32)
+                else: # torch.Tensor
+                    mask = segm
+        self.mask = mask
         self.size = size
         self.mode = mode
 
     def transpose(self, method):
         if method not in (FLIP_LEFT_RIGHT, FLIP_TOP_BOTTOM):
-            raise NotImplementedError(
-                "Only FLIP_LEFT_RIGHT and FLIP_TOP_BOTTOM implemented"
-            )
+            raise NotImplementedError("Only FLIP_LEFT_RIGHT and FLIP_TOP_BOTTOM implemented")
 
         width, height = self.size
         if method == FLIP_LEFT_RIGHT:
-            dim = width
-            idx = 2
+            max_idx = width
+            dim = 1
         elif method == FLIP_TOP_BOTTOM:
-            dim = height
-            idx = 1
+            max_idx = height
+            dim = 0
 
-        flip_idx = list(range(dim)[::-1])
-        flipped_masks = self.masks.index_select(dim, flip_idx)
-        return Mask(flipped_masks, self.size, self.mode)
+        flip_idx = torch.tensor(list(range(max_idx)[::-1]))
+        flipped_mask = self.mask.index_select(dim, flip_idx)
+        return Mask(flipped_mask, self.size, self.mode)
 
     def crop(self, box):
+        box = [int(b) for b in box]
         w, h = box[2] - box[0], box[3] - box[1]
-
-        cropped_masks = self.masks[:, box[1] : box[3], box[0] : box[2]]
-        return Mask(cropped_masks, size=(w, h), mode=self.mode)
+        cropped_mask = self.mask[box[1]: box[3]+1, box[0]: box[2]+1]
```

Comment:
The indexing here depends on the data: we should have a convention to add the +1 or not. I'm undecided about it for the moment. Ideally, training with polygons or masks should give the same result, and I'm unsure this is the case now

Reply:
Cropping on polygons does not need the `w,h`, but I think this is necessary here.

Reply:
Sounds good to me. I will verify this.

Reply:
@fmassa You're right about this. I've verified through tests.